ILM: add searchable snapshot action

andreidan opened the pull request:
Add ILM support for searchable snapshots in the cold phase. Searchable snapshots are part of the lazy snapshot restores effort that's being tracked [here](https://github.com/elastic/elasticsearch/issues/50999)

The API for configuring creating a searchable snapshot is:

```
"cold": {
  "searchable_snapshot" : { 
    "snapshot_repository" : "snapshotRepositoryName"
  }
}
```

andreidan committed:
ILM: add searchable snapshot action

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/ILM+SLM)

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Add license header

andreidan committed:
Document why we don’t restore the lifecycle setting

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RestoreSnapshotStep.java:
```diff
@@ -0,0 +1,80 @@
+ * designated index name.
+ */
+public class RestoreSnapshotStep extends AsyncActionStep {
+    public static final String NAME = "restore-snapshot";
```
```suggestion
    public static final String NAME = "restore-searchable-snapshot";
```

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RestoreSnapshotStep.java:
```diff
@@ -0,0 +1,80 @@
+            }
+
+            @Override
+            public void onFailure(Exception e) {
```
Should we do some special processing for an exception due to a snapshot or restore already being in progress?

andreidan replied:
I thought we'd retry using our usual retry mechanism. Might be good to report we encountered this better though

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -0,0 +1,203 @@
+ * newly created searchable snapshot backed index.
+ */
+public class SearchableSnapshotAction implements LifecycleAction {
+    public static final String NAME = "searchable_snapshot";
```
```suggestion
    public static final String NAME = "searchable-snapshot";
```

dakrone replied:
I think this one may have been lost in some of the other reviews, but I still think it should be changed

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -0,0 +1,203 @@
+    public static final ParseField SNAPSHOT_REPOSITORY = new ParseField("snapshot_repository");
+    public static final ParseField SEARCHABLE_REPOSITORY = new ParseField("searchable_repository");
+
+    public static final String RESTORED_INDEX_PREFIX = "ilm-restored-";
```
We might want to discuss what we want the prefix to be, I'm not sure about "ilm-restored-" in case we ever have a different (non-searchable) ILM action to restore a snapshot, what do you think?

andreidan replied:
Hm, I'm not sure I understand what you mean. The intention was to keep it "searchable snapshot" independent, which it is now. Do you have something specific in mind? (and this is also configurable in the `RestoreSnapshotStep`, in case we reuse it with regulard snapshot restores)

dakrone replied:
What about `restored-`? I think that fits a little more with our `shrink-` prefix

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -0,0 +1,203 @@
+            copyMetadataStep, copySettingsStep, swapAliasesAndDeleteSourceIndexStep);
+    }
+
+    private TriConsumer<Client, IndexMetaData, AsyncWaitStep.Listener> getCheckSnapshotStatusAsyncAction(String snapshotRepository) {
```
Can you add javadocs to this method?

dakrone replied:
And maybe this can be static? We could pass in the names of things, might make it easier to test?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -0,0 +1,203 @@
+                    } else if (snapshotState.equals(ABORTED) || snapshotState.equals(FAILED) || snapshotState.equals(MISSING)) {
+                        listener.onFailure(new SnapshotException(snapshotStatus.getSnapshot(),
+                            "unable to create snapshot [" + snapshotName + "] for index [ " + indexName + "] as part of policy [" +
+                                policyName + "] execution"));
```
Can you capture the `snapshotState` in this message also?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SwapAliasesAndDeleteSourceIndexStep.java:
```diff
@@ -0,0 +1,91 @@
+ * source index.
+ */
+public class SwapAliasesAndDeleteSourceIndexStep extends AsyncActionStep {
+    public static final String NAME = "swap-aliases-to-restored";
```
I think we could probably switch this to just `"swap-aliases"`

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SwapAliasesAndDeleteSourceIndexStep.java:
```diff
@@ -0,0 +1,91 @@
+                .indexRouting(aliasMetaDataToAdd.indexRouting())
+                .searchRouting(aliasMetaDataToAdd.searchRouting())
+                .filter(aliasMetaDataToAdd.filter() == null ? null : aliasMetaDataToAdd.filter().string())
+                .writeIndex(aliasMetaDataToAdd.writeIndex()));
```
There's a comment above this about inheriting everything except for the writeIndex, but this looks like we're inheriting the write index setting?

andreidan replied:
Ah, yes, I'm planning to refactor this into one component and reuse it in ShrinkSetAliasStep. It's not clear to me why we wouldn't want to inherit the write index (as we're deleting the original index in this one update). Do you know why we're skipping it in the shrink case?

dakrone replied:
I don't think it's for a purposeful reason (at least I can't remember it being intentional)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SwapAliasesAndDeleteSourceIndexStep.java:
```diff
@@ -0,0 +1,91 @@
+                .writeIndex(aliasMetaDataToAdd.writeIndex()));
+        });
+
+        getClient().admin().indices().aliases(aliasesRequest, new ActionListener<>() {
```
If you want, in the future, you can use `ActionListener.wrap(resp -> listener.onResponse(true), listener::onFailure)` I believe, rather than a new anonymous class

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SwapAliasesAndDeleteSourceIndexStep.java:
```diff
@@ -0,0 +1,91 @@
+        getClient().admin().indices().aliases(aliasesRequest, new ActionListener<>() {
+
+            @Override
+            public void onResponse(AcknowledgedResponse response) {
```
We should check that the request was actually acknowledged and throw an exception if it wasn't so that it can be retried

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TakeSnapshotStep.java:
```diff
@@ -0,0 +1,66 @@
+
+import static org.elasticsearch.xpack.core.ilm.LifecycleExecutionState.fromIndexMetadata;
+
+public class TakeSnapshotStep extends AsyncActionStep {
```
I think a better name might be `CreateSnapshotStep`

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TakeSnapshotStep.java:
```diff
@@ -0,0 +1,66 @@
+        // expensive operations (ie. clusterStateProcessed)
+        request.waitForCompletion(false);
+        // TODO should we expose this?
+        request.includeGlobalState(false);
```
I don't think we should expose this (meaning to hardcode it as false like you have), these are specific snapshots used only for searchable snapshots

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TakeSnapshotStep.java:
```diff
@@ -0,0 +1,66 @@
+        getClient().admin().cluster().createSnapshot(request, new ActionListener<>() {
+            @Override
+            public void onResponse(CreateSnapshotResponse createSnapshotResponse) {
+                listener.onResponse(true);
```
We should probably check the `status()` of the response, it's possible it could be a failure in which case I think we could retry again immediately?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForGreenIndexHealthStep.java:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Optionally derives the index name using the provided prefix (if any) and waits for the status of the index to be GREEN.
+ */
+public class WaitForGreenIndexHealthStep extends ClusterStateWaitStep {
```
I think we could refactor the users of this class to use `WaitForIndexColorStep` and then we wouldn't need to add this class

andreidan replied:
Totally, thanks for the tip. I didn't know of its existence

dakrone reviewed (commented):
Thanks for working on this @andreidan, I know it's still a WIP, but I left a bunch of comments!

andreidan committed:
Checkstyle

andreidan committed:
Remove unused import

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CleanupSnapshotStep.java:
```diff
@@ -0,0 +1,71 @@
+
+            @Override
+            public void onFailure(Exception e) {
+                if (e instanceof SnapshotMissingException) {
```
Sure

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CopySettingsStep.java:
```diff
@@ -0,0 +1,81 @@
+    public CopySettingsStep(StepKey key, StepKey nextStepKey, Client client, String indexPrefix, String... settingsKeys) {
+        super(key, nextStepKey, client);
+        Objects.nonNull(indexPrefix);
+        Objects.nonNull(settingsKeys);
```
Good catch (I hadn't even acknowledged the existence of an API that just does `obj != null` - TIL :) )

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CopySettingsStep.java:
```diff
@@ -0,0 +1,81 @@
+            settings.put(key, value);
+        }
+
+        UpdateSettingsRequest updateSettingsRequest = new UpdateSettingsRequest(indexName)
```
Oh interesting. I agree it'd be a better avenue

elasticmachine commented:
Pinging @elastic/es-ui (:ES-UI)

andreidan committed:
Use WaitForIndexColorStep

andreidan committed:
Use requireNonNull

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CleanupSnapshotStep.java:
```diff
@@ -0,0 +1,71 @@
+ * Deletes the snapshot designated by the snapshot name present in the lifecycle execution state, hosted in the configured repository.
+ */
+public class CleanupSnapshotStep extends AsyncActionStep {
+    public static final String NAME = "cleanup-snapshot";
```
I think leaving it as "cleanup-snapshot" makes sense as the step is agnostic of searchable snapshots. It issues a "delete snapshot" request for a particular repository and snapshot name (and in our use case we're using it to clean up the regular repository, not the searchable one)

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateSnapshotNameStep.java:
```diff
@@ -0,0 +1,109 @@
+        return generateSnapshotName(name, new ResolverContext());
+    }
+
+    public static String generateSnapshotName(String name, IndexNameExpressionResolver.Context context) {
```
I moved that snapshot generation method here (this is _that_ essentially) as the line of dependency we have is SLM depends on ILM and I didn't want to make it circular.

cjcenizal commented:
@andreidan Would you mind adding some details to the PR description? The ES UI team would be particularly interested in changes to the API and an explanation of the use cases for new features.

andreidan committed:
Refactorings and cleanups

- renamed TakeSnapshotStep to CreateSnapshotStep
- made the Create and RestoreSnapshotStep retry-able using the
AsyncRetryDuringSnapshotActionStep infrastructure
- CreateSnapshotStep checks the response status for internal server
error and fails the step in this case so we retry before we go to the
next steps that’ll discover the snapshot was not created successfully
- CleanupSnapshotStep reports missing repository in a more detailed way
- CopySettingsStep copies the settings using the cluster state directly
- GenerateSnapshotNameStep validates the snapshot name

andreidan committed:
Make performAction final in AsyncRetryDuringSnapshotActionStep

andreidan committed:
Add docs and redesign OnAsyncWaitBranchingStep

This redesign is meant to allow the client to specify that
it wants to stop waiting and move to the “condition unfulfilled”
side of the branch.

andreidan committed:
Rename BranchingStepListener onStopWaitingForCondition

andreidan committed:
Extract and reuse the alias swap request

andreidan committed:
Rename restoredIndexPrefix to targetIndexPrefix

andreidan committed:
Fix tests

andreidan committed:
Unused imports

andreidan committed:
Handle null next step key

andreidan committed:
Try with resources in test

andreidan committed:
Rename originalIndex to sourceIndexName

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
More tests

andreidan committed:
Add more steps tests

andreidan committed:
Make method static

andreidan committed:
Don’t fail if the response isAck flag is false

We’re deleting the original index as part of the alias swap so a
retry of this step after the cluster state change has been committed
will fail

andreidan committed:
Searchable action tests and more step tests

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Store the snapshot name in the ILM execution state

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Remove import

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Copy the repository and snapshot name

andreidan committed:
Expose repo and snapshot name in the ILM explain response

andreidan committed:
Explain lifecycle repo and snapshot in transport

andreidan committed:
Delete action has an optional parameter delege_generated_snapshot

The parameter defaults to true but the users can configure the delete
action to keep the snapshot they generated in the cold phase

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

andreidan committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Drop the searchable repository action param

andreidan committed:
Drop sys.out.println

andreidan committed:
Fix docs tests

andreidan commented:
@elasticmachine elasticsearch-ci/bwc

andreidan commented:
@elasticmachine run elasticsearch-ci/bwc

andreidan committed:
BWC for DeleteAction delete generated snap flag

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Update settings version when copying settings

andreidan committed:
Add test that deletes the snapshot

andreidan committed:
Document step

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/CleanupSnapshotStepTests.java:
```diff
@@ -0,0 +1,156 @@
+                @Override
+                public void onFailure(Exception e) {
+                    fail("expecting the step to report success if repository information is missing from the ILM execution state as there" +
+                        " is no snapshot to deleter");
```
Super minor nit
```suggestion
                        " is no snapshot to delete");
```

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/CleanupSnapshotStepTests.java:
```diff
@@ -0,0 +1,156 @@
+            step.performAction(indexMetaData, clusterState, null, new AsyncActionStep.Listener() {
+                @Override
+                public void onResponse(boolean complete) {
+                }
```
Could probably add `assertTrue(complete)` here

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/CreateSnapshotStepTests.java:
```diff
@@ -0,0 +1,169 @@
+            CreateSnapshotStep step = new CreateSnapshotStep(randomStepKey(), randomStepKey(), client);
+            step.performAction(indexMetaData, clusterState, null, new AsyncActionStep.Listener() {
+                @Override
+                public void onResponse(boolean complete) {
```
Can add `assertTrue(complete)` here as well (and a `fail` clause in the `onFailure` below)

dakrone commented on x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:
```diff
@@ -1545,6 +1548,145 @@ public void testHaltAtEndOfPhase() throws Exception {
         assertBusy(() -> assertFalse("expected " + index + " to be deleted by ILM", indexExists(index)));
     }
 
+    public void testSearchableSnapshotAction() throws Exception {
```
We could probably put these somewhere else to start splitting this apart similar to #53488 (up to you of course)

andreidan replied:
Yes, totally, I am about to start doing that on master (and will backport here with the appropriate conflict fixes) as I think the scope of this PR is already quite large and some infrastructure needs to be extracted from `TimeSeriesLifecycleActionsIT` to be reused across the various new test classes.

dakrone commented on x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java:
```diff
@@ -1545,6 +1548,145 @@ public void testHaltAtEndOfPhase() throws Exception {
+                Map<String, Object> explainIndex = explainIndex(restoredIndexName);
+                String action = (String) explainIndex.get("action");
+                snapshotName[0] = (String) explainIndex.get("snapshot_name");
+                return DeleteAction.NAME.equals(action);
```
Isn't this index going to be deleted by the policy (time value of 0 to delete), so we should probably just check that the index is missing and then make sure the snapshot still exists?

dakrone reviewed (changes requested):
Thanks for working on this Andrei! I left some comments for you!

andreidan committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Format code

Co-Authored-By: Lee Hinman <dakrone@users.noreply.github.com>

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CleanupSnapshotStep.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Override
+    public boolean isRetryable() {
+        return false;
```
I was thinking since we're retrying using the `AsyncRetryDuringSnapshotActionStep` we might not want to make the steps retryable _as well_, but on second thought I think we should (as we would indeed want to retry on rest api call failures for eg). I'll swap this to true.

andreidan committed:
Make CleanupSnapshotStep retryable

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CreateSnapshotStep.java:
```diff
@@ -0,0 +1,72 @@
+        request.waitForCompletion(false);
+        request.includeGlobalState(false);
+        request.masterNodeTimeout(getMasterTimeout(currentClusterState));
+        getClient().admin().cluster().createSnapshot(request,
```
Great catch @dakrone, thank you

andreidan committed:
CreateSnapshotStep request specifies index name

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteAction.java:
```diff
@@ -5,38 +5,63 @@
     public static final String NAME = "delete";
 
-    private static final ObjectParser<DeleteAction, Void> PARSER = new ObjectParser<>(NAME, DeleteAction::new);
+    public static final ParseField DELETE_GENERATED_SNAPSHOT_FIELD = new ParseField("delete_generated_snapshot");
```
I agree. I didn't use this name before because it was sort of tied to the searchable repository, but that's not an issue anymore.

andreidan committed:
Fix typo

Co-Authored-By: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Format code

Co-Authored-By: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Rename delete action field to delete_searchable_snapshot

andreidan committed:
Rename SSA repo field to snapshot_repository (underscore)

andreidan committed:
Fix delete action test assertions

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Fix doc tests

andreidan committed:
Change debug level to warn

andreidan committed:
Log when we stop waiting for the branch condition

andreidan committed:
MountSnapshotStep is retries when response is not acked

andreidan committed:
Doc null handling

andreidan committed:
Update info message to be good english

andreidan committed:
Use the correct Info class Luke

andreidan committed:
Cleanup if/else branches that check the snapshot state

andreidan committed:
Log that the swap aliases response was not acked

andreidan committed:
SnapshotInProgress null check

andreidan commented:
thanks for the review @dakrone I think I've covered all your suggestions

the one thing left to decide is if we should also allow the action in the warm phase. any updates on this @DaveCTurner @tlrx ? (we can add it quite easily later otherwise)

andreidan committed:
Fix test

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan commented:
@elasticmachine run elasticsearch-ci/2

andreidan committed:
Delete phase min_age after 5 seconds

andreidan committed:
Checkstyle

andreidan committed:
Use underscore in action name too

andreidan commented:
@elasticmachin run elasticsearch-ci/2

andreidan commented:
@elasticmachine run elasticsearch-ci/2

andreidan committed:
Get the generated snapshot name from the original index in test

This speeds up the test a bit and also avoids a race condition where the
explain API will not get a response as the ILM is busy executing cluster
action steps for the restored index.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CopySettingsStep.java:
```diff
@@ -0,0 +1,112 @@
+
+        if (settingsKeys == null || settingsKeys.length == 0) {
+            return clusterState;
+        }
```
Super minor, but we can probably stick this before the `targetIndexMetadata == null` check to avoid an error where no settings are copied?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteAction.java:
```diff
@@ -5,38 +5,63 @@
+    private final boolean deleteSearchableSnapshot;
+
     public DeleteAction() {
+        this(true);
```
It might be worth having a separate discussion (I can't remember if we discussed it) with David & Co about whether we want the default for this to be true or false (can be had after this PR)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateSnapshotNameStep.java:
```diff
@@ -0,0 +1,173 @@
+import static org.elasticsearch.xpack.core.ilm.LifecycleExecutionState.fromIndexMetadata;
+
+/**
+ * Generates a snapshot name for the given index and records it in the index metadata along with the provided snapshot repository.
```
Super minor, but can you put an example name in the docs here like "for example, the snapshot will be named '2020-01-02-myindex-mypolicy-randomuuid'"?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MountSnapshotStep.java:
```diff
@@ -0,0 +1,119 @@
+            ActionListener.wrap(response -> {
+                if (response.status() != RestStatus.OK) {
+                    logger.debug("mount snapshot response failed to complete");
+                    throw new ElasticsearchException("mount snapshot response failed to complete");
```
Super minor change:
```suggestion
                    throw new ElasticsearchException("mount snapshot response failed to complete, got response " + response.status());
```

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/OnAsyncWaitBranchingStep.java:
```diff
@@ -0,0 +1,148 @@
+    private SetOnce<Boolean> onCompleteConditionMet;
+
+    /**
+     * {@link BranchingStep} is a step whose next step is based on the {@link #asyncWaitAction} condition being met or not.
```
I think this might be linking to the wrong thing, maybe
```suggestion
     * {@link OnAsyncWaitBranchingStep} is a step whose next step is based on the {@link #asyncWaitAction} condition being met or not.
```

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotInProgressStep.java:
```diff
@@ -0,0 +1,127 @@
+        // we got here before the cluster state was updated with the (or any) snapshot being started, so we'll wait
+        if (snapshotsInProgress == null) {
+            return new Result(false, new Info(String.format(Locale.ROOT,
+                "snapshot [%s] generated by policy [%s] for index [%s] is still in progress", snapshotName, policyName, indexName)));
```
I actually think we need to return `true` here, it's possible that the snapshot was taken and completed (prior to the step being updated), and then the cluster was restarted, so the `SnapshotsInProgress` is null and won't ever exist because the snapshot was successful, best to err on the side of moving forward and doing a check than be wedged I think

andreidan replied:
Deleted the WaitForSnapshotInProgressStep altogether as we'll design a step that waits for a generation change in the repo metadata in a subsequent PR.

dakrone reviewed (changes requested):
Okay whew, sorry about the delay on this Andrei, I left a few more really minor comments, I think the only thing that's major is the `null` check boolean flipping in `WaitForSnapshotInProgressStep.java`

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

andreidan committed:
Update exception message

Co-Authored-By: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Update javadoc

Co-Authored-By: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Add example of generated snapshot name

andreidan committed:
Fail CleanupSnapshotStep if the request is not acked

andreidan committed:
Move empty settings check before target index check

andreidan committed:
Delete WaitForSnapshotInProgressStepTests

We'll add a new way to save snapshot status api calls based on the
repository data generation in a future PR.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'feature/searchable-snapshots' into ilm-searchable-snapshot-action

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/SearchableSnaposhotActionTests.java:
```diff
@@ -0,0 +1,204 @@
+import static org.elasticsearch.xpack.core.ilm.SearchableSnapshotAction.getCheckSnapshotStatusAsyncAction;
+import static org.hamcrest.Matchers.is;
+
+public class SearchableSnaposhotActionTests extends AbstractActionTestCase<SearchableSnapshotAction> {
```
Super minor, but there's a typo here (and in the name of the file):
`SearchableSnaposhotActionTests` vs
`SearchableSnapshotActionTests`

andreidan replied:
Ah, it was hard to spot the typo, great eyes, thanks @dakrone ! 

Regarding the action name, I renamed it to "searchable_snapshot" (lowercase) on purpose when we renamed "snapshot-repository" to "snapshot_repository" as I thought it was confusing to have a configuration that mixes scores and underscores eg.:
```
  "searchable-snapshot" : { 
    "snapshot_repository" : "snapshotRepositoryName"
  }
```
I using underscores everywhere employs less cognitive load: 
```
  "searchable_snapshot" : { 
    "snapshot_repository" : "snapshotRepositoryName"
  }
```

andreidan replied:
The other "composed" action we have is `set_priority` so it would be inconsistent to use scores. If you don't have a strong opinion on this, I think I'd prefer we use underscores everywhere.

dakrone replied:
I read the other one as the step name rather than the action name, I agree that the action name should use underscores (like our other ones), sorry about that!

dakrone reviewed (approved):
LGTM! I left two really minor typo change comments, thanks for iterating so much on this @andreidan!

andreidan committed:
Fix typo

andreidan commented:
Thanks for reviewing this rather huge chunk of functionality Lee
